xla/mlir/backends/openxla/conversion/de_bufferization.h

```diff
@@ -0,0 +1,108 @@
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+#ifndef XLA_MLIR_BACKENDS_OPENXLA_CONVERSION_DE_BUFFERIZATION_H_
+#define XLA_MLIR_BACKENDS_OPENXLA_CONVERSION_DE_BUFFERIZATION_H_
+
+#include "llvm/ADT/DenseMap.h"
+#include "llvm/ADT/SmallVector.h"
+#include "mlir/Dialect/MemRef/IR/MemRef.h"  // from @llvm-project
+#include "mlir/IR/BuiltinTypes.h"  // from @llvm-project
+#include "mlir/IR/Value.h"  // from @llvm-project
+
+namespace xla {
+namespace gpu {
+
+// As a part of the compilation pipeline to prepare XLA executable to run on top
+// of OpenXLA/IREE we convert it from LMHLO dialects to IREEInput dialect. The
+// key difference is that at LMHLO level program operates on buffers (memrefs),
+// and at IREEInput level it uses value semantics and tensors.
+//
+// We rely on in-place semantics of IREEInput operations (tied operands) to
+// re-write operations writing to buffers to operations updating tensors.
+//
+// Example:
+//
+//   func @main(%arg0: memref<f32>, %arg1: memref<f32>) {
+//     lmhlo.fusion {
+//       %0 = bufferization.to_tensor %arg0 : memref<f32>
+//       %0 = bufferization.to_tensor %arg1 : memref<f32>
+//       %2 = mhlo.add %0, %1: tensor<f32>
+//       memref.tensor_store %2, %arg1 : memref<f32>
+//     }
+//     "some.consumer"(%arg1) : (memref<f32>) -> ()
+//   }
+//
+// In this example `%arg0` is a read only buffer, and `%arg1` is a read-write
+// buffer.
+//
+//  func @main(%arg0: tensor<f32>, %arg1: tensor<f32>) {
+//     %0 = iree_input.dispatch @add(%arg0, %arg1)
+//       : (tensor<f32>, tensor<f32>) -> %arg1
+//     "some.consumer"(%0) : (tensor<f32>) -> ()
+//   }
+//
+// We use `DeBufferization` to track the mapping from a memref to the last
+// tensor produced by an operation that wrote into a memref. In the example
+// above instead of passing `%arg1` to the consumer, we pass the last tensor
+// that shares the storage with `%arg1`. After lowering this representation to
+// IREEs HAL dialect, it's guaranteed that XLA program will read/write from/to
+// exactly the same buffer slices as its original version.
+//
+// In XLA all input arguments do not alias, so we don't need any buffer aliasing
+// analysis, and we can safely rely on memref.view operation offsets and sizes.
+//
+// Conversion implementation is a bit more complicated because we have to handle
+// memref.view and memref.reinterpret_cast operations, but all conversions
+// conceptually are doing the same transformation as in example above.
```

Comment:
I love the documentation! Thanks so much for taking great care to it :)

Reply:
Glad someone reads it! :)